stlab/concurrency/wait.hpp

```diff
@@ -0,0 +1,223 @@
+        {
+            std::unique_lock<std::mutex> lock{m};
+            flag = true;
+            condition.notify_one(); // must notify under lock
+        }
+    });
+
+#if STLAB_TASK_SYSTEM(PORTABLE)
+
+    auto resolved = [&] {
+        std::unique_lock<std::mutex> lock{m};
+        return flag;
+    };
+
+    if (!resolved() && !detail::pts().wake() && !resolved()) detail::pts().add_thread();
+
+    /*
+        If no tasks are available we wait for one tick of the system clock and exponentially
+        back off on the wait as long as no tasks are available.
+    */
+
+    for (auto backoff{std::chrono::steady_clock::duration{std::chrono::milliseconds{1}}}; true;
+         backoff *= 2) {
+        {
+            std::unique_lock<std::mutex> lock{m};
+            if (condition.wait_for(lock, backoff, [&] { return flag; })) {
+                break;
+            }
+        }
+        detail::pts().wake(); // try to wake something to unstick.
+    }
+
+#else
+
+    {
+        std::unique_lock<std::mutex> lock{m};
+        condition.wait(lock, [&] { return flag; });
+    }
+
+#endif
+
+    return detail::_get_ready_future<T>{}(std::move(x));
+}
+
+namespace detail {
+
+template <class T>
+struct blocking_get_guarded {
+    bool _timed_out{false};
+    future<T> _result;
+    std::condition_variable _condition;
+    std::mutex _mutex;
+
+    template <class F>
+    auto set(F&& r) -> future<T> {
+        {
+            std::unique_lock<std::mutex> lock{_mutex};
+            if (_timed_out) return std::forward<F>(r);
+            _result = std::forward<F>(r);
+        }
+        _condition.notify_one();
+        /*
+            A valid future<> result is required because the following reduction in
+            recover() will attach a continuation - which will crash on in invalid
+            future<>.
+         */
+        return make_exceptional_future<T>(
+            std::make_exception_ptr(std::runtime_error("not an error")), immediate_executor);
+    }
+
+    auto wait_for(const std::chrono::nanoseconds& timeout) -> future<T> {
+        std::unique_lock<std::mutex> lock{_mutex};
+        _timed_out = !_condition.wait_for(lock, timeout, [&] { return _result.valid(); });
+        return _timed_out ? future<T>{} : std::move(_result);
+    }
+};
+
+} // namespace detail
+
+template <class T>
+auto get_wait_for(future<T> x, const std::chrono::nanoseconds& timeout) -> future<T> {
+    if (x.is_ready()) return x;
+
+#if STLAB_TASK_SYSTEM(PORTABLE)
+    if (!detail::pts().wake()) detail::pts().add_thread();
+#endif
+
+    auto p = std::make_shared<detail::blocking_get_guarded<T>>();
+
+    auto hold = std::move(x).recover(immediate_executor, [_p = stlab::make_weak_ptr(p)](auto&& r) {
+        if (auto p = _p.lock()) return p->set(std::forward<decltype(r)>(r));
+        return std::forward<decltype(r)>(r);
+    });
+
+    auto result = p->wait_for(timeout);
+    return result.valid() ? std::move(result) : std::move(hold);
```

Comment:
As far as I remember, moving out of a function is not recommended and can cause some compilers to emit a warning.

Reply:
It won't work without the moves here - technically this is not moving out of a function, but moving out of the expression. That tossed me for a loop.